### DISCUSSION TITLE
What are your thoughts on Uniform CSS?

### DISCUSSION THREAD

**Discussion by thiagomajesk** [twbs/bootstrap#35014] [General]
I'd like to know, what are your thoughts on [Uniform CSS](https://uniformcss.com), have your head about it?

It seems to me it's like a Tailwind version for Sass with some facilities around CSS generation. 

After skimming the documentation I was wondering... How cool would be to have that integrated into Bootstrap!?
I acknowledge that the current utility API in V5 is good enough for some cases, but it's still somewhat limited in scope.
 
So, do you think that it would be a good idea for Bootstrap V6 to get better integration with a lib like this for CSS generation? My perception is that Bootstrap could use Uniform CSS to regain some terrain over the selling points of Tailwind, which are mostly about 'developer ergonomics'. I'd see this as a complete win for both sides.

PS.: There's nothing that prevents us from using both Uniform and Bootstrap together. But I think that convention/naming-wise, it would be better if there was some intent on improving the default experience for users.

**Comment by ghost** [comment]
I am unable to think of how it enhance user experience to even better?
Can you explain a lill more or refer to that issue?

Plus how about tailwind structure?

**Reply by thiagomajesk** [reply]
Hi @Sahib111! I'm glad I've caught your interest, here's what I'm thinking...

Right now, there are two main popular ways of building pages:

- Using the utility-first approach, which TailwindCSS enables by giving users well-though utilities with a consistent naming convention. After that, if you need to create components, you'll have to `@apply` or abstract those into partials. 

- Using the component-first approach, which Bootstrap enables by giving users common patterns used by most web pages. After that, if you need to customize components, you'll have to override, create a variation, or use utilities to modify styles.

Can you spot the similarities!? Both frameworks empower users in two different ways - I'd even say in complementary ways (_that's why you see people trying to use Tailwind on top of Bootstrap and also why you see people trying to recreate Bootstrap with Tailwind_).

Bootstrap knows how useful utilities and components are and has been slowly adhering to a mixed approach, even more now with the new utility API in V5. The problem is that right now you are kinda [locked-in](https://github.com/twbs/bootstrap/discussions/33885) using the API to generate utilities, if we could relax the conventions a little bit it would be easier to integrate with other tools like Uniform. One good example of that being the current naming conventions that mix component size and breakpoints (eg: `text-md-start`, `shadow-lg`, `btn-lg`).

Now imagine if we had a standard using something like Uniform as an engine for utilities; I think we would get the best of both worlds. That is a collection of battle-tested and quality-proven components with the ability to generate and customize utilities. Bootstrap V5 already tries to provide that but it's not nearly as powerful or extendable as Uniform seems to be.
